File: library/android-legacy/src/androidMain/kotlin/com/adsbynimbus/dynamicprice/Mapping.kt
package com.adsbynimbus.dynamicprice

import com.adsbynimbus.NimbusAd

/** Provides a mapping from a Nimbus response to Dynamic Price target */
fun interface Mapping {
    fun getTarget(ad: NimbusAd): String
}

/** A mapping using a linear step function to generate the target dynamic price value */
typealias LinearPriceGranularity = LinearPriceMapping.Granularity

/**
 * A mapping composed of multiple LinearPriceGranularities in ascending order.
 *
 * Linearly searches the granularity mappings and returns the keywords to be inserted
 */
class LinearPriceMapping(vararg val granularities: Granularity) :
    Mapping, Comparator<LinearPriceMapping.Granularity> {

    init { granularities.sortWith(this) }

    override fun getTarget(ad: NimbusAd): String = granularities.run {
        firstOrNull { ad.bidInCents() < it.max } ?: last()
    }.getTarget(ad)

    override fun compare(o1: Granularity, o2: Granularity): Int = o1.min - o2.min

    /**
     * A mapping using a linear step function to generate the target dynamic price value
     *
     * @param min the minimum line item bid in cents
     * @param max the maximum line item bid in cents
     * @param step the step size, defaults to 20 cents
     */
    class Granularity(val min: Int, val max: Int, val step: Int) : Mapping {

        override fun getTarget(ad: NimbusAd): String =
            (ad.bidInCents() - ad.bidInCents() % step).coerceIn(min, max).toString()
    }
}

File: library/ios/src/main/swift/Mapping.swift
//
//  Mapping.swift
//  
//
//  Created by Jason Sznol on 8/13/26.
//

import NimbusKit

/// Provides a mapping from a Nimbus response to Dynamic Price target
public protocol Mapping {
    func getTarget(_ ad: NimbusAd) -> String
}

/// A mapping using a linear step function to generate the target dynamic price value
public struct LinearPriceGranularity: Comparable, Sendable {

    /// The minimum bid in cents
    public let min: Int

    /// The maximum bid in cents
    public let max: Int

    /// The step size for each line item mapping
    public let step: Int

    /**
     Constructs a new `LinearPriceGranularity`

     - Parameters:
     - min:The minimum bid in cents
     - max: The maximum bid in cents
     - step: The step size for each line item mapping
     */
    public init(min: Int, max: Int, step: Int) {
        self.min = min
        self.max = max
        self.step = step
    }

    /// :nodoc:
    public static func < (lhs: LinearPriceGranularity, rhs: LinearPriceGranularity) -> Bool {
        lhs.min < rhs.min
    }
}

/// A mapping composed of multiple LinearPriceGranularities in ascending order
public struct LinearPriceMapping: Mapping, Sendable {

    /// The granularities used in this mapping
    public let granularities: [LinearPriceGranularity]

    /**
     Constructs a new `LinearPriceMapping`

     - Parameters:
     -  granularities: the granularities to use
     */
    public init(granularities: [LinearPriceGranularity]) {
        self.granularities = granularities.sorted()
    }

    /// Constructs a new `LinearPriceMapping`
    public init(_ granularities: LinearPriceGranularity...) {
        self.init(granularities: granularities)
    }

    /**
     Linearly searches the granularity mappings and returns the keywords to be inserted

     - Parameters:
     - ad: An ad from Nimbus

     - Returns: The keywords to set
     */
    public func getTarget(_ ad: NimbusAd) -> String {
        let range = granularities.first { ad.bidInCents < $0.max } ?? granularities.last!
        return String(min(max(ad.bidInCents - (ad.bidInCents % range.step), range.min), range.max))
    }
}
